rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadata.java

```diff
@@ -1,154 +1,215 @@
-      throw new IllegalArgumentException(
-          "composite metadata entry buffer is too short to contain proper entry");
+  /** An {@link Entry} backed by an explicitly declared MIME type. */
+  public static final class ExplicitMimeTimeEntry implements Entry {
+
+    private final ByteBuf content;
+
+    private final String type;
+
+    public ExplicitMimeTimeEntry(ByteBuf content, String type) {
+      this.content = content;
+      this.type = type;
     }
-    if (decoded == CompositeMetadataFlyweight.METADATA_BUFFERS_DONE) {
-      throw new NoSuchElementException("composite metadata has no more entries");
+
+    @Override
+    public ByteBuf getContent() {
+      return this.content;
     }
 
-    ByteBuf header = decoded[0];
-    this.content = decoded[1];
-    // move the nextEntryIndex
-    this.nextEntryIndex =
-        CompositeMetadataFlyweight.computeNextEntryIndex(this.nextEntryIndex, header, content);
-
-    if (header.readableBytes() == 1) {
-      this.id = CompositeMetadataFlyweight.decodeMimeIdFromMimeBuffer(header);
-      WellKnownMimeType wkn = WellKnownMimeType.fromId(id);
-      if (wkn != WellKnownMimeType.UNKNOWN_RESERVED_MIME_TYPE) {
-        this.mime = wkn.getMime();
-      }
-    } else {
-      this.id = -1;
-      CharSequence charSequence = CompositeMetadataFlyweight.decodeMimeTypeFromMimeBuffer(header);
-      if (charSequence == null) {
-        throw new IllegalArgumentException(
-            "composite metadata entry parsing failed on custom type");
-      }
-      this.mime = charSequence.toString();
+    @Override
+    public String getMimeType() {
+      return this.type;
     }
   }
 
   /**
-   * Returns the mime type {@link String} representation of the currently decoded entry, if there is
-   * one.
-   *
-   * <p>A {@code null} value should only occur with a positive {@link #getCurrentMimeId()}, denoting
-   * an entry that is compressed but unparseable (probably a buffer encoded by another version of
-   * the well known mime type extension spec, which is only reserved in this implementation).
-   *
-   * @return the mime type for this entry, or null
+   * An {@link Entry} backed by a {@link WellKnownMimeType} entry, but one that is not understood by
+   * this implementation.
    */
-  @Nullable
-  public String getCurrentMimeType() {
-    return this.mime;
-  }
+  public static final class ReservedMimeTypeEntry implements Entry {
+    private final ByteBuf content;
+    private final int type;
 
-  /**
-   * @return the compressed mime id byte if relevant (0-127), or -1 if not (custom mime type as
-   *     {@link String})
-   */
-  public byte getCurrentMimeId() {
-    return this.id;
+    public ReservedMimeTypeEntry(ByteBuf content, int type) {
+      this.content = content;
+      this.type = type;
+    }
+
+    @Override
+    public ByteBuf getContent() {
+      return this.content;
+    }
+
+    @Override
+    public String getMimeType() {
```

Comment:
should we override the javadoc to state that this is _the_ case where the getter returns `null`?